pallets/rad-claims/src/lib.rs

```diff
@@ -0,0 +1,585 @@
+        /// can be used for malicious spams or Deny of Service (DoS) attacks.
+        fn validate_unsigned(
+            _source: TransactionSource, 
+            call: &Self::Call
+        ) -> TransactionValidity {
+            if let Call::claim(account_id, amount, sorted_hashes) = call {
+                // Check that proofs are valid with a root that exists in the root hash storage
+                if Self::verify_proofs(account_id, amount, sorted_hashes.into()) {
+                    return ValidTransaction::with_tag_prefix("RadClaims")
+                        .priority(T::UnsignedPriority::get())
+                        .and_provides((account_id, amount, sorted_hashes))
+                        .longevity(TryInto::<u64>::try_into(
+                            T::Longevity::get())
+                            .unwrap_or(64_u64))
+                        .propagate(true)
+                        .build()
+                } else {
+                    return InvalidTransaction::BadProof.into();
+                }
+            }
+    
+            InvalidTransaction::Call.into()
+        }
+    }
+} // end of 'pallet' module
+
+
+// ----------------------------------------------------------------------------
+// Pallet implementation block
+// ----------------------------------------------------------------------------
+
+// Pallet implementation block.
+//
+// This main implementation block contains two categories of functions, namely:
+// - Public functions: These are functions that are `pub` and generally fall into 
+//   inspector functions that do not write to storage and operation functions that do.
+// - Private functions: These are private helpers or utilities that cannot be called 
+//   from other pallets.
+impl<T: Config> Pallet<T> {
+   
+    /// Return the account identifier of the crowdloan claim pallet.
+	///
+	/// This actually does computation. If you need to keep using it, then make
+	/// sure you cache the value and only call this once.
+	pub fn account_id() -> T::AccountId {
+	  T::PalletId::get().into_account()
+	}
+
+    /// TODO: need function documentation
+    /// Hash a:b if a < b, else b:a. Uses the runtime module's hasher.
+    pub fn sorted_hash_of(a: &T::Hash, b: &T::Hash) -> T::Hash {
+        let mut h: Vec<u8> = Vec::with_capacity(64);
+        if a < b {
+            h.extend_from_slice(a.as_ref());
+            h.extend_from_slice(b.as_ref());
+        } else {
+            h.extend_from_slice(b.as_ref());
+            h.extend_from_slice(a.as_ref());
+        }
+
+        T::Hashing::hash(&h).into()
+    }
+
+    /// Returns true if the given origin can update the upload account
+    fn can_update_upload_account(origin: T::Origin) -> DispatchResult {
+        T::AdminOrigin::try_origin(origin)
+            .map(|_| ())
+            .or_else(ensure_root)?;
+
+        Ok(())
+    }
+
+    // TODO: need function documentation
```

Comment:
Can we do this in the same refactoring? Or remove this maybe.

Reply:
Yes. It will be nice if Miguel or Ved could explain this function.

Reply:
Essentially this verifies proofs that have been lexicographically sorted when building a tree.
1. Form a leaf Hash(encoded(account_id+amount))
2. Pass leaf to iterator as the first accumulative value to the sorted_hash_of function
3. sorted_hash_of hashes both hash1 and hash2 together, and the order depends on which one is "bigger"
4. This approach avoids having an extra byte that tells if the hash is left or right so they can be concatenated accordingly before hashing
5. Then check that the resulting root hash matches with the one stored